src/scores/contingency.py

```diff
@@ -0,0 +1,258 @@
+CategoryTable.
+'''
+
+import operator
+import xarray
+
+DEFAULT_PRECISION = 8
+
+def make_binary_table(proximity, match_operator):
+	table = match_operator.make_table(proximity)
+	return table
+
+def make_category_table(proximity, category_operator):
+	table = category_operator.make_table(proximity)
+	return table
+
+class BinaryContingencyTable():
+	'''
+	At each location, the value will either be:
+	 - A true positive    (0)
+	 - A false positive   (1)
+	 - A true negative    (2)
+	 - A false negative   (3)
+
+	It will be common to want to operate on masks of these values,
+	such as:
+	 - Plotting these attributes on a map
+	 - Calculating the total number of these attributes
+	 - Calculating various ratios of these attributes, potentially
+	   masked by geographical area (e.g. accuracy in a region)
+
+	As such, the per-pixel information is useful as well as the overall
+	ratios involved.
+	'''
+
+	def __init__(self, forecast_events, observed_events):
+
+		# type checking goes here
+		self.forecast_events = forecast_events
+		self.observed_events = observed_events
+
+		self.tp = (self.forecast_events == 1) & (self.observed_events == 1)  # true positives
+		self.tn = (self.forecast_events == 0) & (self.observed_events == 0)  # true negatives
+		self.fp = (self.forecast_events == 1) & (self.observed_events == 0)  # false positives
+		self.fn = (self.forecast_events == 0) & (self.observed_events == 1)  # false negatives
+
+		# Variables for count-based metrics, calculated on first access
+		self.tp_count = self.tp.sum().values.item()     # Count of true positives
+		self.tn_count = self.tn.sum().values.item()     # Count of true negatives
+		self.fp_count = self.fp.sum().values.item()     # Count of false positives
+		self.fn_count = self.fn.sum().values.item()     # Count of true negatives
+		self.total_count = self.tp_count + self.tn_count + self.fp_count + self.fn_count
+
+	def accuracy(self) -> float:
+		'''
+		The proportion of forecasts which are true
+		'''
+
+		correct_count = self.tp_count + self.tn_count
+		ratio = correct_count / self.total_count
+		return ratio
+
+	def frequency_bias(self) -> float:
+		'''
+		'''
+		freq_bias = (self.tp_count + self.fp_count) / (self.tp_count + self.fn_count)
+		return freq_bias
+
+	def probability_of_detection(self) -> float:
+		'''
+		What proportion of the observed events where correctly forecast?
+		Range: 0 to 1.  Perfect score: 1.
+		'''
+
+		pod = self.tp_count / (self.tp_count + self.fn_count)
+		return pod
+
+	def false_alarm_rate(self) -> float:
+		'''
+		What fraction of the non-events were incorrectly predicted?
+		Range: 0 to 1.  Perfect score: 0.
+		'''
+
+		far = self.fp_count / (self.tn_count + self.fp_count)
+		return far
+
+	def success_ratio(self) -> float:
+		'''
+		What proportion of the forecast events actually eventuated?
+		Range: 0 to 1.  Perfect score: 1.
+		'''
+
+		sr = self.tp_count / (self.tp_count + self.fp_count)
+		return sr
+
+	def threat_score(self) -> float:
```

Comment:
This is more commonly called CSI in the met literature. Consider creating an alias for this function. This could be done for other metrics that have multiple names.